File: src/julia/kamada_kawai/algo/seq.jl
module Seq

export seq

using ...Kamada_Kawai

function derivative_x_m_seq(kk::KamadaKawai, index::Int)
    sum = 0.0
    for i in 1:kk.n
        if i != index
            dist_x = kk.coords[index].x - kk.coords[i].x
            dist_y = kk.coords[index].y - kk.coords[i].y

            sum += kk.k_ij[index, i] * (dist_x - ((kk.l_ij[index, i] * dist_x) / (dist_x^2 + dist_y^2)^(Float64(0.5))))
        end # if
    end # for
    
    return sum
end # derivative_x_m_seq

function derivative_y_m_seq(kk::KamadaKawai, index::Int)
    sum = 0.0
    for i in 1:kk.n
        if i != index
            dist_x = kk.coords[index].x - kk.coords[i].x
            dist_y = kk.coords[index].y - kk.coords[i].y

            sum += kk.k_ij[index, i] * (dist_y - ((kk.l_ij[index, i] * dist_y) / (dist_x^2 + dist_y^2)^(Float64(0.5))))
        end # if
    end # for
    
    return sum
end # derivative_y_m_seq

function derivative_xx_m_seq(kk::KamadaKawai, index::Int)
    sum = 0.0
    for i in 1:kk.n
        if i != index
            dist_x = kk.coords[index].x - kk.coords[i].x
            dist_y = kk.coords[index].y - kk.coords[i].y

            sum += kk.k_ij[index, i] * (1 - (kk.l_ij[index, i] * dist_y^2) / (dist_x^2 + dist_y^2)^(Float64(1.5)))
        end # if
    end # for
    
    return sum
end # derivative_xx_m_seq

function derivative_yy_m_seq(kk::KamadaKawai, index::Int)
    sum = 0.0
    for i in 1:kk.n
        if i != index
            dist_x = kk.coords[index].x - kk.coords[i].x
            dist_y = kk.coords[index].y - kk.coords[i].y

            sum += kk.k_ij[index, i] * (1 - (kk.l_ij[index, i] * dist_x^2) / (dist_x^2 + dist_y^2)^(Float64(1.5)))
        end # if
    end # for
    
    return sum
end # derivative_yy_m_seq

function derivative_xy_m_seq(kk::KamadaKawai, index::Int)
    sum = 0.0
    for i in 1:kk.n
        if i != index
            dist_x = kk.coords[index].x - kk.coords[i].x
            dist_y = kk.coords[index].y - kk.coords[i].y

            sum += kk.k_ij[index, i] * (kk.l_ij[index, i] * dist_x * dist_y) / (dist_x^2 + dist_y^2)^(Float64(1.5))
        end # if
    end # for
    
    return sum
end # derivative_xy_m_seq

function get_delta_y_seq(
    derivative_x_m_seq::Float64,
    derivative_y_m_seq::Float64,
    derivative_xx_m_seq::Float64,
    derivative_yy_m_seq::Float64,
    derivative_xy_m_seq::Float64
)
    return (-(derivative_xy_m_seq * derivative_x_m_seq) + (derivative_xx_m_seq * derivative_y_m_seq)) / (-(derivative_xx_m_seq * derivative_yy_m_seq) + derivative_xy_m_seq^2)
end # get_delta_y_seq

function get_delta_x_seq(
    derivative_y_m_seq::Float64,
    derivative_yy_m_seq::Float64,
    derivative_xy_m_seq::Float64,
    delta_y::Float64
)
    return (-(derivative_y_m_seq) - (derivative_yy_m_seq * delta_y)) / derivative_xy_m_seq
end # get_delta_x_seq

function delta_m_seq(derivative_x_m_seq::Float64, derivative_y_m_seq::Float64)
    return (derivative_x_m_seq^2 + derivative_y_m_seq^2)^Float64(0.5)
end # delta_m_seq

function calculate_delta_seq(kk::KamadaKawai, index::Int)
    derivative_x = derivative_x_m_seq(kk, index)
    derivative_y = derivative_y_m_seq(kk, index)

    return delta_m_seq(derivative_x, derivative_y)
end # calculate_delta_seq

function calculate_deltas_seq(kk::KamadaKawai)
    deltas = Vector{Float64}(undef, kk.n)

    for i in 1:kk.n
        deltas[i] = calculate_delta_seq(kk, i)
    end # for

    return deltas
end # calculate_deltas_seq

function get_max_delta_m_index_seq(deltas::Vector{Float64}, epsilon::Float64)
    max = -1.0
    index = -1

    for i in 1:length(deltas)
        delta = deltas[i]
        if delta > max
            max = delta
            if delta > epsilon
                index = i
            end # if
        end # if
    end # for

    return index
end # get_max_delta_m_index_seq

function seq(kk::KamadaKawai)
    states = Vector{Vector{Coord}}()
    # copy the coordinates into the first element of states
    push!(states, deepcopy(kk.coords))

    deltas = calculate_deltas_seq(kk)
    max_delta_m_seq_index = get_max_delta_m_index_seq(deltas, kk.epsilon)

    while max_delta_m_seq_index != -1
        while deltas[max_delta_m_seq_index] > kk.epsilon
            d_x_m = derivative_x_m_seq(kk, max_delta_m_seq_index)
            d_y_m = derivative_y_m_seq(kk, max_delta_m_seq_index)
            d_xx_m = derivative_xx_m_seq(kk, max_delta_m_seq_index)
            d_yy_m = derivative_yy_m_seq(kk, max_delta_m_seq_index)
            d_xy_m = derivative_xy_m_seq(kk, max_delta_m_seq_index)

            delta_y = get_delta_y_seq(d_x_m, d_y_m, d_xx_m, d_yy_m, d_xy_m)
            delta_x = get_delta_x_seq(d_y_m, d_yy_m, d_xy_m, delta_y)

            kk.coords[max_delta_m_seq_index].x += delta_x
            kk.coords[max_delta_m_seq_index].y += delta_y

            deltas[max_delta_m_seq_index] = calculate_delta_seq(kk, max_delta_m_seq_index)
        end # while
        
        deltas = calculate_deltas_seq(kk)
        max_delta_m_seq_index = get_max_delta_m_index_seq(deltas, kk.epsilon)
        
    end # while
    push!(states, deepcopy(kk.coords))

    return states
end # seq

end # module

File: src/go/kamada_kawai/algo/seq.go
package algo

import (
	"math"
)

func get_delta_y_seq(
	derivative_x_m float64,
	derivative_y_m float64,
	derivative_xx_m float64,
	derivative_yy_m float64,
	derivative_xy_m float64) float64 {
	return (-(derivative_xy_m * derivative_x_m) + (derivative_xx_m * derivative_y_m)) / (-(derivative_xx_m * derivative_yy_m) + (derivative_xy_m * derivative_xy_m))
}

func get_delta_x_seq(
	derivative_y_m float64,
	derivative_yy_m float64,
	derivative_xy_m float64,
	delta_y float64) float64 {
	return (-(derivative_y_m) - (derivative_yy_m * delta_y)) / (derivative_xy_m)
}

func delta_m_seq(derivative_x float64, derivative_y float64) float64 {
	return math.Sqrt(math.Pow(derivative_x, 2) + math.Pow(derivative_y, 2))
}

func calculate_delta_seq(kk *KamadaKawai, index int) float64 {
	derivative_x := kk.derivative_x_m(index)
	derivative_y := kk.derivative_y_m(index)

	return delta_m_seq(derivative_x, derivative_y)
}

func calculate_deltas_seq(kk *KamadaKawai) []float64 {
	var deltas []float64 = make([]float64, kk.N)

	for i := 0; i < kk.N; i++ {
		deltas[i] = calculate_delta_seq(kk, i)
	}

	return deltas
}

func get_max_delta_m_index_seq(deltas []float64, epsilon float64) int {
	var max float64 = -1
	var index int = -1

	for i, delta := range deltas {
		if delta > max {
			max = delta
			if delta > epsilon {
				index = i
			}
		}
	}

	return index
}

func (kk *KamadaKawai) Seq() [][]Coord {

	states := make([][]Coord, 1)
	states[0] = make([]Coord, kk.N)
	copy(states[0], kk.Coords)

	var deltas []float64 = calculate_deltas_seq(kk)
	var max_delta_m_seq_index int = get_max_delta_m_index_seq(deltas, kk.Epsilon)

	for {
		if max_delta_m_seq_index == -1 {
			break
		}

		for {
			var d_x_m = kk.derivative_x_m(max_delta_m_seq_index)
			var d_y_m = kk.derivative_y_m(max_delta_m_seq_index)
			var d_xx_m = kk.derivative_xx_m(max_delta_m_seq_index)
			var d_yy_m = kk.derivative_yy_m(max_delta_m_seq_index)
			var d_xy_m = kk.derivative_xy_m(max_delta_m_seq_index)

			var delta_y float64 = get_delta_y_seq(d_x_m, d_y_m, d_xx_m, d_yy_m, d_xy_m)
			var delta_x float64 = get_delta_x_seq(d_y_m, d_yy_m, d_xy_m, delta_y)

			kk.Coords[max_delta_m_seq_index].X += delta_x
			kk.Coords[max_delta_m_seq_index].Y += delta_y

			deltas[max_delta_m_seq_index] = calculate_delta_seq(kk, max_delta_m_seq_index)
			if deltas[max_delta_m_seq_index] < kk.Epsilon {
				break
			}
		}

		deltas = calculate_deltas_seq(kk)
		max_delta_m_seq_index = get_max_delta_m_index_seq(deltas, kk.Epsilon)
	}
	state := make([]Coord, kk.N)
	copy(state, kk.Coords)
	states = append(states, state)

	return states
}
